### PR TITLE
Add Granite 4.1 Vision as built-in multimodal model

### PR DESCRIPTION
**Pull request by artem-spector**
## Summary

- Add built-in support for **Granite 4.1 Vision** (`granite4_vision`), a multimodal model with deepstack vision-to-LLM feature injection and window Q-Former downsampling
- Model uses SigLip2 vision encoder, injects features at multiple LLM layers via masked addition (deepstack), with optional spatial offset sampling
- Includes `_CONFIG_REGISTRY` entry so vLLM can load the model without upstream HF transformers support
- Model ships with a same-repo LoRA adapter; tests use `default_mm_loras` for automatic adapter activation

### Files added/modified

| File | Purpose |
|---|---|
| `vllm/model_executor/models/granite4_vision.py` | Model implementation (deepstack injection, multimodal interface) |
| `vllm/model_executor/models/registry.py` | Register in `_MULTIMODAL_MODELS` |
| `vllm/transformers_utils/configs/granite4_vision.py` | Config class for `_CONFIG_REGISTRY` |
| `vllm/transformers_utils/configs/__init__.py` | Config import |
| `vllm/transformers_utils/config.py` | `_CONFIG_REGISTRY` entry |
| `vllm/transformers_utils/processors/granite4_vision.py` | Processor class |
| `vllm/transformers_utils/processors/__init__.py` | Processor import |
| `docs/models/supported_models.md` | Add to supported models list |
| `tests/models/registry.py` | Registry test entry |
| `tests/models/multimodal/generation/test_common.py` | VLM correctness test (HF vs vLLM) |

## Test plan

- [x] `pytest tests/models/test_registry.py -k "Granite4Vision"` — model loading (1 passed)
- [x] `pytest tests/models/multimodal/generation/test_common.py -k "granite4_vision"` — HF vs vLLM output comparison with LoRA adapter (1 passed)
- [x] `pytest tests/models/multimodal/processing/test_common.py -k "ibm-granite/granite-vision-4.1-4b"` — multi-modal processing correctness (3 passed)

### CODE REVIEW

File: tests/models/multimodal/generation/test_common.py [hunk 1]
```diff
@@ -425,6 +445,20 @@
         auto_cls=AutoModelForImageTextToText,
         vllm_output_post_proc=model_utils.llava_image_vllm_to_hf_output,
     ),
+    "granite4_vision": VLMTestInfo(
```
Comment (DarkLight1337) [inline tests/models/multimodal/generation/test_common.py]: Please keep in alphabetical order

File: vllm/model_executor/models/granite4_vision.py [hunk 2]
```diff
@@ -0,0 +1,988 @@
...
+import os
+from collections.abc import Iterable, Mapping
+from fractions import Fraction
+
+import torch
+import torch.nn as nn
+from safetensors.torch import load_file
+from transformers import BatchFeature
+from transformers.models.blip_2.configuration_blip_2 import Blip2QFormerConfig
+from transformers.models.blip_2.modeling_blip_2 import Blip2QFormerModel
```
Comment (DarkLight1337) [inline vllm/model_executor/models/granite4_vision.py]: You can use the `Blip2QFormerModel` from vLLM

File: vllm/model_executor/models/granite4_vision.py [hunk 3]
```diff
@@ -0,0 +1,988 @@
...
+            # encoder output list).
+            self.vision_tower = SiglipVisionModel(
+                config.vision_config,
+                quant_config=quant_config,
+                require_post_norm=False,
+                prefix=maybe_prefix(prefix, "vision_tower"),
+            )
+
+            # image_newline parameter
+            if getattr(config, "use_image_newline_parameter", False):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/granite4_vision.py]: Can we just access `config.use_image_newline_parameter` directly without `getattr`?

File: vllm/model_executor/models/granite4_vision.py [hunk 4]
```diff
@@ -0,0 +1,988 @@
...
+
+            # Deepstack projectors: one per (vision_layer, llm_layer) pair
+            self.layerwise_projectors = nn.ModuleList([
+                WindowQFormerDownsampler(config)
+                for _ in range(len(config.deepstack_layer_map))
+            ])
+
+            # Spatial projectors: 4 offset groups
+            self.spatial_projectors = None
+            if getattr(config, "use_spatial_sampling", False):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/granite4_vision.py]: Same

File: vllm/model_executor/models/granite4_vision.py [hunk 5]
```diff
@@ -0,0 +1,988 @@
...
+        ds_map: dict[int, torch.Tensor] = {}
+        for llm_layer_idx, features in self._ds_features:
+            ds_map[llm_layer_idx] = features
+
+        vision_mask = self._ds_vision_mask
+        if vision_mask is not None:
+            vision_mask = vision_mask[:num_tokens]
+
+        # Run through decoder layers with deepstack injection
+        for i in range(lm_inner.start_layer, lm_inner.end_layer):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/granite4_vision.py]: Can we handle this like in Qwen3-VL where a separate `LLMModel` class is defined?

**Review by DarkLight1337 — COMMENTED** [review]
Thanks, some initial comments

File: tests/models/multimodal/generation/test_common.py (hunk 1 above)
Comment (artem-spector) [reply tests/models/multimodal/generation/test_common.py]: Done in a92e14d.

File: vllm/model_executor/models/granite4_vision.py (hunk 2 above)
Comment (artem-spector) [reply vllm/model_executor/models/granite4_vision.py]: Done in https://github.com/vllm-project/vllm/commit/b3ac3860f067ef10ca492374c1c366e03ee7a899 — switched WindowQFormerDownsampler to use vLLM's Blip2QFormerModel from blip2.py

File: vllm/model_executor/models/granite4_vision.py (hunk 3 above)
Comment (artem-spector) [reply vllm/model_executor/models/granite4_vision.py]: Done in a92e14d.

File: vllm/model_executor/models/granite4_vision.py (hunk 4 above)
Comment (artem-spector) [reply vllm/model_executor/models/granite4_vision.py]: Done in a92e14d.

File: vllm/model_executor/models/granite4_vision.py (hunk 5 above)
Comment (artem-spector) [reply vllm/model_executor/models/granite4_vision.py]: Done in 35bb790 — added Granite4VisionLLMModel (subclass of GraniteModel) and Granite4VisionLLMForCausalLM (subclass of GraniteForCausalLM). 
The deepstack layer loop now lives inside Granite4VisionLLMModel.forward() with a deepstack_input_embeds: IntermediateTensors | None parameter, matching the Qwen3-VL pattern.

File: vllm/model_executor/models/granite4_vision.py [hunk 6]
```diff
@@ -0,0 +1,988 @@
...
+            layer = lm_inner.layers[i]
+
+            # Inject deepstack features at target layers (before layer forward)
+            if i in ds_map and vision_mask is not None and vision_mask.any():
+                features = ds_map[i][:num_tokens]
+                hidden_states = hidden_states.clone()
+                hidden_states[vision_mask] = (
+                    hidden_states[vision_mask] + features[vision_mask]
+                )
```
[the remark this answers is not in the document]
Comment (artem-spector) [reply vllm/model_executor/models/granite4_vision.py]: Fixed in b26fb94 — replaced clone+assign with in-place addition.

File: vllm/model_executor/models/granite4_vision.py [hunk 7]
```diff
@@ -0,0 +1,988 @@
...
+                N, hidden_size,
+                dtype=inputs_embeds.dtype,
+                device=inputs_embeds.device,
+            )
+            buf[is_multimodal] = concat_features.to(dtype=inputs_embeds.dtype)
+            prepared.append((llm_layer, buf))
+
+        self._ds_features = prepared
+        self._ds_vision_mask = is_multimodal
+        self._ds_level_features = []  # consumed
```
[the remark this answers is not in the document]
Comment (artem-spector) [reply vllm/model_executor/models/granite4_vision.py]: Fixed in 35bb790 — self._ds_features and self._ds_vision_mask are eliminated entirely. 
Vision features are now packed into a wide tensor in embed_multimodal() (cached by the framework), unpacked in embed_input_ids() into pre-allocated persistent GPU buffers (one per deepstack level, fixed address), and passed to the layer loop as IntermediateTensors. 
The buffers are zeroed after each forward pass. This matches the Qwen3-VL pattern and is safe under CUDA graph capture and chunked prefill.

File: vllm/model_executor/models/granite4_vision.py [hunk 8]
```diff
@@ -0,0 +1,1079 @@
...
+                            (0, 0, 0, max(0, num_tokens - buf_len)),
+                        )
+                    hidden_states = hidden_states + feat
+            hidden_states = layer(positions, hidden_states)
+
+        if not get_pp_group().is_last_rank:
+            return IntermediateTensors({"hidden_states": hidden_states})
+
+        hidden_states = self.norm(hidden_states)
+        return hidden_states
```
[the remark this answers is not in the document]
Comment (artem-spector) [reply vllm/model_executor/models/granite4_vision.py:296]: Fixed in commit 453479aaf.                                                                                                                                                                                                                                                                                                                                                                                                                                                                 

The fix has three parts:                                                                                                                                                                                                                                                                                                                                                                                                                                                               
1. make_empty_intermediate_tensors on the right class — overridden on Granite4VisionLLMForCausalLM (not on Granite4VisionLLMModel), which is the class vLLM actually calls. It adds a ds_{layer} key for each deepstack target layer so the  
PP receive buffer is pre-allocated with the right shape.                                                                                                                                                                                   
2. _ds_layer_indices accessible to the override — the inner LLM model only sees text_config (no deepstack_layer_map), so the outer model sets self.language_model._ds_layer_indices after construction.                                      
3. Full-size buffers sent between ranks — sync_and_slice_intermediate_tensors copies copy_len = num_tokens_padded rows (CUDA graph bucket size), which can exceed actual token count. Sending the full pre-allocated buffer (max_tokens × hidden_size) ensures the copy always succeeds.

File: vllm/model_executor/models/granite4_vision.py [hunk 9]
```diff
@@ -0,0 +1,1079 @@
...
+        if buf0.device != inputs_embeds.device or buf0.dtype != inputs_embeds.dtype:
+            self._ds_buffers = [
+                b.to(device=inputs_embeds.device, dtype=inputs_embeds.dtype)
+                for b in self._ds_buffers
+            ]
+
+        for level_idx in range(len(self._ds_layer_indices)):
+            buf_data = torch.zeros(N, lm_h, dtype=inputs_embeds.dtype, device=inputs_embeds.device)
+            buf_data[is_multimodal] = level_features[level_idx]
+            self._ds_buffers[level_idx][:N].copy_(buf_data)
```
[the remark this answers is not in the document]
Comment (artem-spector) [reply vllm/model_executor/models/granite4_vision.py]: Fixed in 53c0b8c. Eliminated the per-call torch.zeros allocation — now zeroing the persistent buffer slice directly and scattering features into it:
```
target = self._ds_buffers[level_idx][:N]
target.zero_()
target[is_multimodal] = level_features[level_idx]

```
On the register_buffer suggestion: since _ds_buffers is a list of tensors rather than named attributes, register_buffer doesn't apply directly. The lazy device migration on first use is safe because text-only requests never reach embed_input_ids (they skip embed_multimodal entirely), so CUDA graph capture won't encounter CPU tensors.

File: tests/models/multimodal/generation/test_common.py [hunk 10]
```diff
@@ -86,6 +86,26 @@
...
+
+    Self-contained to avoid calling AutoConfig/AutoTokenizer without
+    trust_remote_code (needed while the model is not in upstream HF).
+    """
+    output_ids, output_str, out_logprobs = vllm_output
+    mm_token_id = 100352
+    hf_output_ids = [
+        token_id
+        for idx, token_id in enumerate(output_ids)
+        if token_id != mm_token_id or output_ids[idx - 1] != mm_token_id
```
[the remark this answers is not in the document]
Comment (artem-spector) [reply tests/models/multimodal/generation/test_common.py]: Fixed in 53c0b8c. Added the explicit idx == 0 guard:
`
if token_id != mm_token_id or idx == 0 or output_ids[idx - 1] != mm_token_id                                                                                                                                                                 
`

**Review by DarkLight1337 — COMMENTED** [review]
Can you add this model to the offline inference vision examples? e.g. https://github.com/vllm-project/vllm/blob/main/examples/offline_inference/vision_language_multi_image.py

**Review by DarkLight1337 — APPROVED** [review]
Thanks for your patience!

File: vllm/model_executor/models/granite4_vision.py [hunk 11]
```diff
@@ -0,0 +1,988 @@
...
+                continue
+            A = lora_a[module_key].to(merge_device).float()
+            B = lora_b[module_key].to(merge_device).float()
+            delta = scaling * (B @ A)
+            if _add_delta(module_key + ".weight", delta):
+                merged += 1
+            else:
+                logger.warning("LoRA target not found: %s", module_key)
+
+        return merged
```
[the remark this answers is not in the document]
Comment (artem-spector) [reply vllm/model_executor/models/granite4_vision.py]: Resolved in bb6a9636 — dropped the entire merge-on-load flow. Native vLLM LoRA serving (--enable-lora --default-mm-loras) is the supported path for LoRA adapters.

File: vllm/model_executor/models/granite4_vision.py [hunk 12]
```diff
@@ -0,0 +1,1079 @@
...
+                if hasattr(module, "_get_shard_offset_mapping"):
+                    # QKVParallelLinear: string shard_id ("q", "k", "v")
+                    shard_offset = module._get_shard_offset_mapping(shard_id)
+                    if shard_offset is not None:
+                        shard_size = delta.shape[0] // tp_size
+                        tp_delta = delta.narrow(
+                            0, tp_rank * shard_size, shard_size)
+                        shard = param.data[shard_offset:shard_offset + shard_size]
+                        param.data[shard_offset:shard_offset + shard_size] = (
+                            shard.float() + tp_delta.to(shard.device)
```
[the remark this answers is not in the document]
Comment (artem-spector) [reply vllm/model_executor/models/granite4_vision.py]: Resolved in bb6a9636 — dropped the entire merge-on-load flow. Native vLLM LoRA serving (--enable-lora --default-mm-loras) is the supported path for LoRA adapters.

**Comment by artem-spector** [comment]
CI tests fail because the model ibm-granite/granite-vision-4.1-4b is not public yet.
We could give an access for your CI user to the model, would that be acceptable?

**Comment by DarkLight1337** [comment]
Let's just set `is_available_online=False` until it's released
